cell_classification/augmentation_pipeline.py

```diff
@@ -130,10 +133,373 @@ def py_aug(batch, tf_aug):
+        """
+        Args:
+            prob (float):
+            The probability of applying the augmentation
+        min_factor (float):
+            The minimum contrast factor
+        max_factor (float):
+            The maximum contrast factor
+        """
+        super(LinearContrast, self).__init__()
+        self.prob = prob
+        self.min_factor = min_factor
+        self.max_factor = max_factor
+
+    def __call__(self, image, labels):
+        """Apply linear contrast augmentation
+        Args:
+            image (tf.Tensor):
+                The image to augment
+                labels (tf.Tensor):
+                The labels to augment
+        Returns:
+            tf.Tensor:
+                The augmented image
+            tf.Tensor:
+                The augmented labels
+        """
+        if tf.random.uniform(()) < self.prob:
+            factor = tf.random.uniform((), self.min_factor, self.max_factor)
+            image *= factor
+        return image, labels
+
+
+class MixUp(tf.Module):
+    """MixUp augmentation for segmentation data"""
+
+    def __init__(self, prob=0.5, alpha=0.2):
+        """
+        Args:
+            prob (float):
+                The probability of applying the augmentation
+            alpha (float):
+                The alpha parameter of the beta distribution
+        """
+        super(MixUp, self).__init__()
+        self.prob = prob
+        self.alpha = alpha
+
+    def __call__(self, x_mplex, x_binary, labels, loss_mask):
+        """Apply mixup augmentation
+        Args:
+            x_mplex (tf.Tensor):
+                The batch of images to augment
+            x_binary (tf.Tensor):
+                The batch of binary images to augment
+            labels (tf.Tensor):
+                The batch of labels to augment
+        Returns:
+            tf.Tensor:
+                The augmented image batch
+            tf.Tensor:
+                The augmented label batch
+        """
+        # cast to float
+        loss_mask = tf.cast(loss_mask, tf.float32)
+        labels = tf.cast(labels, tf.float32)
+        x_binary = tf.cast(x_binary, tf.float32)
+        # calculate mixup coefficient beta and prob [0,1] if an image gets transformed
+        b = tf.shape(x_mplex)[0]
+        prob = tf.random.uniform([b], 0, 1) < self.prob
+        beta = tf.random.uniform([b], 0, 1)
+        beta = tf.math.maximum(beta, 1 - beta)
+        beta = tf.math.pow(beta, 1 / self.alpha)
+        beta = tf.where(prob, beta, tf.ones_like(beta))
+        prob = tf.cast(tf.reshape(prob, [b, 1, 1, 1]), tf.float32)
+        # apply mixup on data
+        beta = tf.reshape(beta, [b, 1, 1, 1])
+        x_mplex = beta * x_mplex + (1 - beta) * tf.reverse(x_mplex, axis=[0])
+        x_binary = beta * x_binary + (1 - beta) * tf.reverse(x_binary, axis=[0])
+        labels = beta * labels + (1 - beta) * tf.reverse(labels, axis=[0])
+        # loss mask should be 1 where both mixup loss masks == 1 and zero otherwise
```

Comment:
Is this the new way you're handling 2s? How come both == 1 is the correct approach for loss masks?

Reply:
Would you rather do the promix weighted sum over the two loss_masks? I didn't know how to approach this and I thought the conservative solution is just masking out both areas, but then a lot of stuff could be masked out.. might be more clever to have to have it weighted then..